### ISSUE TITLE
CC0057: Should not try to remove parameters meant to fulfill a delegate contract

### ISSUE THREAD

**Issue by Haacked** [code-cracker/code-cracker#646]
CC0057 appears to be a bit too aggressive. For example, suppose I need to pass a delegate to a method and the delegate signature has two parameters, but my implementation of the delegate only needs one parameter. Here's a demo class.

```
public class BugDemo
{
    public delegate void ConfigFileChanged(string key, object value);

    public static void FireHandler(ConfigFileChanged changeHandler)
    {
        changeHandler?.Invoke("key", "object");
    }

    public static void Init()
    {
        FireHandler(OnConfigFileChanged);
    }

    static void OnConfigFileChanged(string key, object value)
    {
        // We only care about dependency changes
        if (key == "none")
            return;
        Console.WriteLine(key);
    }
}
```

Code Cracker tells me that the `object value` parameter of `OnConfigFileChanged` is not used and suggests I change the signature. I can't always change the signature because it might be defined elsewhere. In this case, it shouldn't try to remove the parameter. Perhaps it shouldn't even warn in this case.

**Comment by giggio** [comment]
@Haacked Ok, I confirmed the problem. And it is crashing when the code fix would be applied. So, at least that has to be fixed.

We have a few options here on how to fix this:

1. We could update the `ConfigFileChanged` delegate (if it is on source), removing `object value` from the signature and also update the `OnConfigFileChanged`, doing the same thing.
<br />1a. If the delegate comes from an outside reference, we do not raise `CC0057`.
<br />1b. If the delegate is being used somewhere else we do not raise `CC0057`.
2. We could always not raise `CC0057.`

The first option is complicated to work using the current Roslyn APIs, unless we provide some poor implementation. The problem is looking for references is async, and we are not supposed to do async work on the analyzer. The other option would be to collect the references during the compilation and only raise the diagnostic at the end, but this would present other problems.

The second options also suffers from the same problem, as we also would have to look for usage of the `OnConfigFileChanged` (which is async and we can't do).

So, this is not a simple problem. I will take this one and see what I can come up with.

**Comment by Haacked** [comment]
> 1.We could update the  ConfigFileChanged  delegate (if it is on source), removing  object value  from the signature and also update the  OnConfigFileChanged , doing the same thing. 

I'm not sure that's the right thing. The delegate could be defined elsewhere. Or it might be designed appropriately, and only this instance doesn't need parameters. I think it's one thing when you define a method yourself and don't use a parameter. I think it's another thing when you're fulfilling a contract.

For example, it doesn't seem like you raise `CC0057` when you're overriding a virtual method and you don't use all the parameters. I think this is the same situation as that.

**Comment by giggio** [comment]
Well, a delegate is just a method contract... if it is defined on your source, this means you own it. If no one else is using one of it's parameters, why not simply remove the parameter from the delegate and from the method?
But as I said before, that might not even be possible. :)
So I am looking for a way to do not raise it on this situation. Let's see if I can find it.

**Comment by Haacked** [comment]
> Well, a delegate is just a method contract... if it is defined on your source, this means you own it. If no one else is using one of it's parameters, why not simply remove the parameter from the delegate and from the method?

Sure, if it's a private delegate and _every_ usage omits one of the parameters. In the code sample I posted, the delegate is public. Which means there could be callers not under my control who are implementing it. Removing the parameter would break them.

> So I am looking for a way to do not raise it on this situation. Let's see if I can find it.

:+1:

**Comment by giggio** [comment]
@Haacked 
Ok, Here is what I have found. We are going to have to compromise.
We can't use the `SymbolFinder` class for several reasons. The main one being that we cannot use the `Microsoft.CodeAnalysis.Workspaces.dll` during analysis, only when a code fix provider is running. So, we are not able to verify who is referencing the method. On your example, the `Init` method is in the same class, but it could be anywhere else, if `OnConfigFileChanged` was public. There are 2 ways to find this usage:

1. We use the `SymbolFinder`, which we can't.
2. We run on compilation start and compilation end, and we collect any usages as we find them.

Number 2 is not an option also, as it creates problems. For example, it might raise a diagnostic when a compilation ends, and later you change the code, but until the next compilation starts and ends, the diagnostic would stay there on your error list. It could be a while, several seconds, even minutes. You don't want that. This kind of diagnostic only works for command line compilations or for hidden diagnostics, which are only there for code fix providers to pick up, and then you can run all the async stuff and have access to the workspaces dll.

The other reason not to use `SymbolFinder` is that all its methods are async. We can't run async code during code analysis, we can't await, and if we `.Wait` we can cause serious problems.

So, here comes the compromise:

If the method is private, we can search through the class declaration for it's usages, and if we find that the method is being used as a method group, we don't raise the diagnostic. But this will not work with partial classes, as the method might be used on another document, and we will still end up with a false positive, we will raise an incorrect diagnostic. On the code fix, if we find that it is was a partial class and the method was used and the delegate is on source, we could either update the delegate or not provide a code action. I prefer to offer it and let the user decide. If the delegate is not in our source, we simply do not offer a code fix.
If the method is public, we ignore all that. We always offer the code fix and don't bother. Or should we never offer it? I see a lot of benefit to offer it, but I would like to hear your opinion (and other people's opinions as well).

Other options:

1. We could ignore methods that start with `OnXXX`.
2. We could ignore methods on partial classes.

What do you think?

**Comment by Haacked** [comment]
> If the method is private, we can search through the class declaration for it's usages, and if we find that the method is being used as a method group, we don't raise the diagnostic.

Seems legit to me.

I don't mind the code fix being offered. What I don't want to do is litter my code with `#pragmas` to hide this diagnostic in cases where I can't fix it. So not raising the diagnostic when it's a public delegate seems like a good compromise.

**Comment by giggio** [comment]
@Haacked It is finished. Would you try the latest release on the myget feed to make sure it works for you?

**Comment by Haacked** [comment]
@giggio I'll try and get to it soon.

**Comment by Haacked** [comment]
@giggio I saw the diagnostic raised, but it didn't ask me to remove the unused parameter.

**Comment by giggio** [comment]
@Haacked It should not even raise a diagnostic. Can you isolate the code and post it?
RC6 is out just now with the fix, can you try with it?
Also, sometimes you have to close VS after you update the analyzers, it seems that the new ones might not be picked up immediately and diagnostics updated.

**Comment by Haacked** [comment]
@giggio ah! Let me try again. I might have failed to update properly. I think I was on RC5

**Comment by Haacked** [comment]
Everything works great! Thanks!
